app/src/main/java/com/example/cinemapp/ui/main/DetailsViewModel.kt

```diff
@@ -0,0 +1,69 @@
+package com.example.cinemapp.ui.main
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.example.cinemapp.data.MovieRepository
+import com.example.cinemapp.util.MovieUtil
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+
+class DetailsViewModel(
+    private val movieRepository: MovieRepository
+) : ViewModel() {
+
+    data class State(
+        val details: MovieDetails? = null,
+        val cast: List<CastMember> = emptyList(),
+        val media: List<Media> = emptyList()
+    )
+
+    private val _state = MutableStateFlow(State())
+    val state: StateFlow<State> = _state.asStateFlow()
+
+    fun getMovieDetails(movieId: Int) {
+        viewModelScope.launch {
+
+            var newDetails: MovieDetails? = null
+            var newCredits: List<CastMember> = emptyList()
+            var newMedia: List<Media> = emptyList()
+            var newTrailer: Media.Video? = null
+
+            viewModelScope.launch {
+                newDetails = movieRepository.getMovieDetails(movieId)?.let { details ->
```

Comment:
Seems like you are calling 4 different calls inside the same coroutine, making them non paralel...
[See this link](https://stackoverflow.com/a/62966043)